ui/src/app/shared/debounce.ts

```diff
@@ -0,0 +1,21 @@
+export default function debounce<T extends (...args: any[]) => any>(fn: T, debounceMs: number) {
+    let timer: number | null = null;
+
+    const cancel = () => {
+        if (timer !== null) {
+            clearTimeout(timer);
+            timer = null;
+        }
+    };
+
+    const debouncedFn = (...args: Parameters<T>) => {
+        cancel();
+
+        timer = window.setTimeout(() => {
```

Comment:
Why is this timer needed?

Reply:
See the above comment discussing what debounce does (I can add comments to the code as well) - essentially the function that's passed in will not be invoked until the debounced function has not been invoked for N milliseconds